### PR TITLE
[WIP][AccessEnforcementOpts] Add mergeAccesses analysis + optimization

### PR DESCRIPTION
**Pull request by shajrawi**
We have the following code:
```
  %30 = begin_access [read] [dynamic] [no_nested_conflict] %1 : $*UInt64 // users: %31, %32
  %31 = load %30 : $*UInt64                       // user: %34
  end_access %30 : $*UInt64                       // id: %32
  %33 = struct $UInt64 (%27 : $Builtin.Int64)     // user: %34
  %34 = apply %14(%31, %33) : $@convention(thin) (UInt64, UInt64) -> UInt64 // user: %36
  %35 = begin_access [modify] [dynamic] [no_nested_conflict] %1 : $*UInt64 // users: %36, %37
  store %34 to %35 : $*UInt64                     // id: %36
  end_access %35 : $*UInt64                       // id: %37
 ```
If modify access fails then it is inclusive of the read access.

We can merge the two to a bigger “modify” block and reduce the amount of runtime overhead / calls.

We can also do the same for two merge accesses inside the same region, creating one larger scope access, as long as nothing conflicts in the larger scope / prevents us from doing so.

Now consider what happens when we couple what's described above with LICM: we can hoist the newly-merged large-scope access outside of loops! (see attached test case)

I have benchmarked this optimization on `RangeIteration`  from the benchmark suite: it improves performance by over 6X !!!

I will test it on more benchmarks later.

radar rdar://problem/40376124

### CODE REVIEW

**Comment by shajrawi** [comment]
@atrick can you please review?

**Comment by shajrawi** [comment]
@swift-ci Please test

**Comment by shajrawi** [comment]
@swift-ci Please clean test OS X

**Comment by shajrawi** [comment]
Comparing exclusivity enabled on master v.s. this PR:

TEST | OLD | NEW | DELTA | SPEEDUP
-- | -- | -- | -- | --
RangeIterationSigned | 687 | 92 | -86.6% | **7.47x**
DropFirstCountableRangeLazy | 28 | 14 | -50.0% | **2.00x**
PolymorphicCalls | 20 | 14 | -30.0% | **1.43x**
Sim2DArray | 330 | 246 | -25.5% | **1.34x**
ExclusivityGlobal | 4 | 3 | -25.0% | **1.33x**
Chars | 605 | 493 | -18.5% | **1.23x**

**Comment by shajrawi** [comment]
@swift-ci Please test

**Comment by shajrawi** [comment]
@swift-ci Please test

**Comment by jrose-apple** [comment]
Can you go into a little more detail about why this is safe? What if `%14` modifies `%1` through another access path?

**Comment by shajrawi** [comment]
See `FunctionAccessedStorage` analysis (`AccessedStorageAnalysis.h` line 127) : the output of that analysis is checking if there are any unidentified accesses in that function and its callees (`hasUnidentifiedAccess`) and, if we are able to analyze the entire chain and all the accessed storage locations, we have a summary of all the storage locations accessed.
Given `%1`, we have to call the conservative `mayConflictWith` method on `%14` which will tell us if this is safe or not.

**Review by atrick — CHANGES_REQUESTED** [review]
Originally, this was meant to be a natural extension of the optimization that folds nested access. Now it's trying to handle other cases that don't naturally fall out. I think we need to make this much simpler to justify the approach vs. merging through code motion.

File: lib/SILOptimizer/Transforms/AccessEnforcementOpts.cpp [hunk 1]
```diff
@@ -619,6 +623,593 @@ removeLocalNonNestedAccess(const AccessConflictAnalysis::Result &result,
...
+}
+
+// Region Access summary info
+class RegionAccessSummary {
+  const AccessConflictAnalysis::Result &ACAResult;
+  /// Tracks the accesses at the end of each region, for the purpose of
+  /// finding nested conflicts.
+  /// For each storage loaation, maintains only a single entry.
+  using AccessUnordedSet = llvm::SmallDenseSet<BeginAccessInst *>;
+  using RegionToAccessesMap = llvm::SmallDenseMap<unsigned, AccessUnordedSet>;
```
[the remark this answers is not in the document]
Comment (shajrawi) [reply lib/SILOptimizer/Transforms/AccessEnforcementOpts.cpp]: The regions are numbered in `LoopRegionAnalysis`, the unsigned is the region ID / number based on that. I will better explain it.

**Comment by shajrawi** [comment]
replaced by https://github.com/apple/swift/pull/18438 - closing this PR

**Pull request closed without merging**